### PR TITLE
Issue #5553: Jacoco workaround for MainTest

### PR DESCRIPTION
**Pull request by pbludov**
Issue #5553
Workaround for Jacoco limitations (part 3).

### CODE REVIEW

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 1]
```diff
@@ -1061,4 +1061,13 @@ public void testModuleNameInMultiThreadMode() throws Exception {
         }
     }
 
+    @Test
+    public void testJacocoWorkaround() throws Exception {
+        // This test verifies nothing. It is a workaround for the Jacoco limitations.
+        // see https://github.com/jacoco/jacoco/issues/117 for more details.
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Please move this to javadoc like we define in https://github.com/checkstyle/checkstyle/issues/4970 and add a breif description of what the issue is in jacoco.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 2]
```diff
@@ -1061,4 +1061,13 @@ public void testModuleNameInMultiThreadMode() throws Exception {
         }
     }
 
+    @Test
+    public void testJacocoWorkaround() throws Exception {
+        // This test verifies nothing. It is a workaround for the Jacoco limitations.
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: > This test verifies nothing.
> verifyStatic(System.class, times(1));

This statement looks false.

Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: done.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 3]
```diff
@@ -1061,4 +1061,18 @@ public void testModuleNameInMultiThreadMode() throws Exception {
         }
     }
 
+    /**
+     * This test verifies nothing. It is a workaround for the Jacoco limitations.
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @pbludov My original comment wasn't for removing the verify in the method.
I also mentioned that the reason for the mocking should be the first sentence.

Why can't we verify some type of result?
Can we atleast verify `System.exit` was called?

Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: done.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 4]
```diff
@@ -1061,4 +1061,20 @@ public void testModuleNameInMultiThreadMode() throws Exception {
...
+     *
+     * @see <a href="https://github.com/jacoco/jacoco/issues/117">Jacoco issue 117</a>
+     * @throws Exception if error occurs
+     */
+    @Test
+    public void testJacocoWorkaround() throws Exception {
+        mockStatic(System.class);
+        Main.main();
+        verifyStatic(System.class, times(1));
+        System.exit(-1);
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: I don't understand why `System.exit` is inside the test.

Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: So Powermock does validate calls to static methods.
A terrible syntax, but there is no alternative.

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @pbludov I am even more confused.
The point of the `System.exit` verify was to verify the call to that method in production code.
If we are verifying the call in test code, then it is completely pointless and should probably just be removed. The point of tests is to verify production code, not test code.

There is no way to verify `System.exit` is called in production?

Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @rnveach yes it is confusing. The test verifies a call to `System.exit` in _production_ code.
To prove this, just comment out the line https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/Main.java#L241
The test will fail. Today i'll send a patch related to the Mockito2 API. You'll se a lot of such code.

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: So why is this needed in test code then?

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 5]
```diff
@@ -1061,4 +1061,21 @@ public void testModuleNameInMultiThreadMode() throws Exception {
...
     }
 
+    /**
+     * This test is a workaround for the Jacoco limitations.
+     * A call to {@link System#exit(int)} will never return, so Jacoco coverage probe will be
+     * missing. Mocking the {@code System} class will turn {@code System.exit()} to noop.
+     *
+     * @throws Exception if error occurs
+     * @see <a href="https://github.com/jacoco/jacoco/issues/117">Jacoco issue 117</a>
+     * @noinspection CallToSystemExit
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: As I told before, all suppression should be explained in javadoc comment. Please do.

Base on conversation with rnveach, bellow, just make a summary, some link to documentation might help to understand this syntax.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 4 above)
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @pbludov This has still not been resolved.

Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: done. The code reverted to the initial version.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 5 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Done. Actually, there is no reason to check the call to `System.exit`. In other tests, it is already tested with `ExpectedSystemExit`.

**Comment by pbludov** [comment]
Now it fails with "JUnit test method testJacocoWorkaround() contains no assertions"

**Comment by romani** [comment]
@pbludov ,

Please recheck execution of shippable
PROFILE="pitest-checkstyle-common" on your local, it looks like hanging on shippable side

**Comment by pbludov** [comment]
I'm able to reproduce this hung. It's an OOM issue actually. I'm looking for a fix right now.

**Comment by pbludov** [comment]
Note that MainTest is not included into the profile pitest-checkstyle-common. So this PR is not related to this hung.

**Comment by pbludov** [comment]
It seems that the profile pitest-checkstyle-common is too large and should be split.
At one point there are four minions, two of them tries to eat more than 4GB of memory.

**Comment by pbludov** [comment]
Reverting till 8.8.1 release didn't help. The profile is still broken. Tomorrow I will bisect it.

**Comment by romani** [comment]
> Note that MainTest is not included into the profile pitest-checkstyle-common. So this PR is not related to this hung.

Strange that is become problem only with your changes 

I relaunched master build
https://app.shippable.com/github/checkstyle/checkstyle/runs/6387/summary/console to reproduce problem in case some changes were done on CI side.

**Comment by pbludov** [comment]
I sent a bunch of small PRs to resolve this hung. Please do the review.

**Comment by pbludov** [comment]
I think I found the bad one. It is the `PackageNamesLoader` test. Removing `PackageNamesLoader` from the profile `pitest-checkstyle-common` solves the `RUN_ERROR` error.

**Comment by pbludov** [comment]
I think there is some sort of miscommunication between Pitest and Powermock.
Both manipulate the classes at the bytecode level and at some point something somewhere is broken.
It is necessary to isolate all tests that use `mockStatic`, such as `PackageNamesLoaderTest`, to minimize the consequences.

**Comment by pbludov** [comment]
The PR is green now and ready to be merged.

**Pull request merged**